lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs

```diff
@@ -619,58 +878,348 @@ genMakeChangeData = flip suchThat isValidMakeChangeData $ do
+    -> Property
+prop_makeChangeForUnknownAsset_sum weights (asset, quantities) =
+    F.fold quantities === F.fold ((`TokenMap.getQuantity` asset) <$> changes)
+  where
+    changes = makeChangeForUnknownAsset weights (asset, quantities)
+
+prop_makeChangeForUnknownAsset_length
+    :: NonEmpty TokenMap
+    -> (AssetId, NonEmpty TokenQuantity)
+    -> Property
+prop_makeChangeForUnknownAsset_length weights surplus =
+    F.length changes === F.length weights
+  where
+    changes = makeChangeForUnknownAsset weights surplus
+
+unit_makeChangeForUnknownAsset
+    :: [Expectation]
+unit_makeChangeForUnknownAsset =
+    [ makeChangeForUnknownAsset weights surplus `shouldBe` expectation
+    | (weights, surplus, expectation) <- matrix
+    ]
+  where
+    matrix =
+        [ ( m [(assetA, q 1)] :| [m [(assetB, q 1)]]
+          , (assetC, q <$> 1 :| [1])
+          , m [(assetC, q 1)] :| [m [(assetC, q 1)]]
+          )
+
+        , ( m [(assetA, q 1)] :| [m [(assetB, q 1)]]
+          , (assetC, q <$> 1 :| [1, 1])
+          , m [(assetC, q 1)] :| [m [(assetC, q 2)]]
+          )
+
+        , ( m [(assetA, q 1)] :| [m [(assetB, q 1)]]
+          , (assetC, q <$> 1 :| [])
+          , m [(assetC, q 0)] :| [m [(assetC, q 1)]]
+          )
+        ]
+
+    q :: Natural -> TokenQuantity
+    q = TokenQuantity
+
+    m :: [(AssetId, TokenQuantity)] -> TokenMap
+    m = TokenMap.fromFlatList
+
+    assetA :: AssetId
+    assetA = AssetId (UnsafeTokenPolicyId $ Hash "A") (UnsafeTokenName "1")
+
+    assetB :: AssetId
+    assetB = AssetId (UnsafeTokenPolicyId $ Hash "B") (UnsafeTokenName "")
+
+    assetC :: AssetId
+    assetC = AssetId (UnsafeTokenPolicyId $ Hash "A") (UnsafeTokenName "2")
 
 --------------------------------------------------------------------------------
--- Making change for surplus assets
+-- Making change for known assets
 --------------------------------------------------------------------------------
 
-prop_makeChangeForSurplusAssets_length
-    :: Map AssetId (NonEmpty TokenQuantity) -> NonEmpty () -> Property
-prop_makeChangeForSurplusAssets_length assetQuantities target =
-    NE.length (makeChangeForSurplusAssets assetQuantities target)
-        === NE.length target
+prop_makeChangeForKnownAsset_sum
+    :: NonEmpty TokenMap
+    -> (AssetId, TokenQuantity)
+    -> Property
+prop_makeChangeForKnownAsset_sum weights (asset, quantity) =
+    if any (`TokenMap.hasQuantity` asset) weights then
+        quantity === totalChangeValue
+    else
+        totalChangeValue === TokenQuantity 0
+  where
+    changes = makeChangeForKnownAsset weights (asset, quantity)
+    totalChangeValue = F.fold ((`TokenMap.getQuantity` asset) <$> changes)
 
-prop_makeChangeForSurplusAssets_sort
-    :: Map AssetId (NonEmpty TokenQuantity) -> NonEmpty () -> Property
-prop_makeChangeForSurplusAssets_sort assetQuantities target = property $
-    inAscendingPartialOrder (makeChangeForSurplusAssets assetQuantities target)
```

Comment:
Are we sure want to remove this property?

The original motivation for this property was to:

- check that we are limiting the production of change bundles where large quantities of one asset are bundled with tiny quantities of some other asset, and
- to the greatest extent generate bundles where small quantities are bundled with other small quantities and large quantities with other large quantities.

The original comment for `makeChangeForSurplusAssets` did describe this goal, but alas was not clear enough at describing the motivation, which is my oversight!

Reply:
I indeed failed to see the motivation for that property in the first place. Thanks for clearing that up. Sounds like a reasonable motivation, so I can see how to re-introduce that :+1:

Reply:
Great! I think it should be fairly easy to re-introduce. (I suspect that the function still satisfies the property, as this behaviour is provided by `padCoalesce`.)

Reply:
On second thought, I fail a bit to see how the current implementation helps to satisfy the motivation you just highlighted and also, whether this is even something we do want in practice. The current implementation does the following:

```hs
    change :: NonEmpty TokenBundle
    change
        = TokenBundle.fromCoin <$> changeForCoin
        & NE.zipWith (<>) (TokenBundle.fromTokenMap <$> changeForPaymentAssets)
        -- Here we sort in ascending order of coin value so that surplus assets
        -- and coin values are in increasing order in the resulting bundles:
        & NE.sortWith TokenBundle.getCoin
        & NE.zipWith (<>) (TokenBundle.fromTokenMap <$> changeForSurplusAssets)
```

So, it starts off by assigning all surplus coins, and zip that with the change of user-defined assets. After this step, we have in principle change outputs which are (ideally) similar to user-specified outputs; or at least that's what the algorithm is trying to achieve. 

Then, we have a set of assets that aren't present in user-specified and have to distribute them somehow. So the most reasonable thing to do is to leave them untouched as much as possible and merge them with other token bundles. The only real area of maneuver we have is with which other change outputs to merge them? This new rework doesn't make any decision whereas the current did first sort existing tokens by coin quantity so that largest bundles will be merged with largest coin values. I don't quite see why we would generally want that. In practice, UTxOs with the largest coin values will likely be ada-only UTxOs, and it's probably best to keep them as ada-only. 

I do agree with the following motivation though:

> to the greatest extent possible generate bundles where small quantities are bundled with other small quantities and large quantities with other large quantities. 

Besides, with the new implementation, change coins are assigned only after the other changes have been calculated (for we need them to have been created _before_ so that the minimum coin value can be calculated for each). This makes it impossible to re-order the list before change coins are assigned (otherwise, we may deviate from the requested output distribution). 

So instead, since we know that: 

a) `padCoalesce` returns a list ordered in ascending order
b) We can re-order requested outputs for change calculation upfront, so that "bigger" outputs are first 

We can do a best effort at merging the largest non user-defined change assets with the largest user-defined change assets. Another issue is that we only have a partial order on token maps whereas we would need a total order here. It is maybe possible to define one for that purpose, such that:

- We consider the partial order for elements which can be partially ordered. 
- For others, we may consider their maximum quantity. 

e.g.

```
>>> sort [ [ (A,3) ], [ (A,1), (B,2) ], [ (C,4), (B,2) ], [ (B,1) ] ]
[ [ (B,1) ], [ (A,1), (B,2) ], [ (A,3) ], [ (C,4), (B,2) ]
```

Since the algorithm tends to lead to the elimination of small quantities, this should over time help generating UTxOs where large quantities are bundled together and small quantities together.

Reply:
Hi @KtorZ 

I do agree with your reasoning here.

With regard to ordering maps according to their greatest token quantity: I think this is a very reasonable approach.

I was curious, however, as to whether the `totalOrder` function defined in f58841bfdf036f164fb5d82f643a3d48f5a9e9c6 (the one that gives priority to `PartialOrd`) actually produces a well-defined order. So just to make sure, I extracted the `totalOrder` function out, and tested it with `testLaws @Ord`. This revealed a few subtle issues with the ordering.

I did try to fix these issues, but was unable to produce an ordering that gives priority to `PartialOrd` and yet satisfies the `Ord` laws.

However, I do think that we could adopt a simpler approach here:

It could be enough to simply order token maps by their greatest quantity. In the case where there is a tie-break, we can fall back to lexicographic ordering, which provides fast termination and guarantees to produce a total order. (This is similar to the technique we use for ordering `TxOut`.)

To that end, I pushed a commit 185a712b41c91b2203685044b2f9d246efd2fcb4, which fixes the ordering, and also adds a property test to check that the ordering is well-defined, satisfying the `Ord` laws.

I hope this is okay. But if you still feel there is a better ordering, then perhaps we can adjust the `Ord` instance, or go for something else entirely. Since we now have a property test, it should be easy to check that whatever ordering we choose is valid.